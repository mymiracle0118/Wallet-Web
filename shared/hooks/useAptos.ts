import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { Ed25519Keypair } from '@mysten/sui.js/keypairs/ed25519'
import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getAptosWalletUsingSeed, getBalance } from '@portal/shared/services/aptosService'
import { ethers } from 'ethers'
import { toUTF8Array } from '../utils/utf8'

type APTOSWalletState = {
  /** APTOS network - create wallet */
  createAptosWallet: (mnemonics?: string) => Promise<any>
  /** APTOS network - Update balance */
  refreshBalance: (aptosType?: string) => Promise<void>
  _keypair?: Ed25519Keypair | null
  /** APTOS network - Wallet Address */
  address?: string
  /** APTOS network - privateKey Address */
  privateKey?: string
  /** APTOS network - Total Balance of wallet */
  balance: number
  /** APTOS network - All objects owned by wallet */
  objectsOwned?: Array<any>

  wallet?: ethers.Wallet
  encryptedPrivateKey: (string | number)[]
  encryptedWallet?: string
}

const initialState = {
  address: undefined,
  balance: 0,
  _keypair: undefined,
  privateKey: undefined,
  encryptedPrivateKey: [],
}

export const useAptos = create<APTOSWalletState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,

        createAptosWallet: async (mnemonics?: string) => {
          const wallet = await getAptosWalletUsingSeed(mnemonics)
          if (wallet) {
            const encryptedPrivateKey = toUTF8Array(wallet?.privateKey)
            set({ address: wallet?.address, encryptedPrivateKey })
          }

          const { refreshBalance } = get()
          await refreshBalance()
          return wallet
        },
        refreshBalance: async (aptosType?: string) => {
          const { address } = get()
          if (!address) return
          const balance = await getBalance(address, aptosType)
          set({ balance: balance })
        },
      }),
      {
        name: 'APTOS-storage',
        getStorage: () => chromeSyncStorage,
        partialize: (state) => Object.fromEntries(Object.entries(state).filter(([key]) => !['wallet'].includes(key))), // Prevents decrypted wallet from being persisted in storage
      }
    ),
    { name: 'APTOS - Wallet' }
  )
)
