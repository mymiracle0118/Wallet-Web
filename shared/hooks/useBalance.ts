import { useEffect, useMemo } from 'react'
import { useQuery } from '@tanstack/react-query'
import { BigNumber } from 'ethers'

import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

type BalanceResponse = {
  balance: BigNumber | number
  balanceFormatted: number
  isLoading: boolean
}
const { updateTokenBalance } = useStore.getState()
export const useBalance = (token: NetworkToken): BalanceResponse => {
  let balance, balanceFormatted

  const networkFactory = NetworkFactory.selectByNetworkId(token.shortName)
  if (!networkFactory) {
    return {
      balance: 0,
      balanceFormatted: 0,
      isLoading: false,
    }
  }

  const walletAddress = token.address
  const { data, isLoading } = useQuery(
    ['asset-balance', token.networkName, walletAddress, token.shortName],
    () => networkFactory.getTokenBalance(walletAddress as string, token),
    {
      refetchInterval: 60000,
      staleTime: 30000,
    }
  )

  balance = useMemo(() => {
    return data || BigNumber.from(0)
  }, [data])
  balanceFormatted = useMemo(() => networkFactory.getBalanceFormattedValue(data), [data]) as number

  useEffect(() => {
    // Update asset balance in wallet to keep running total
    if (balance && balanceFormatted) {
      updateTokenBalance(token.shortName, balance, balanceFormatted)
    }
  }, [balance, balanceFormatted])

  return {
    balance,
    balanceFormatted,
    isLoading,
  }
}
