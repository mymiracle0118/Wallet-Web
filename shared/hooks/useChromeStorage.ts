import { useCallback, useEffect, useState } from 'react'

export const storage = {
  get: (key: string, defaultValue: any, storageArea: Storage) => {
    const keyObj = defaultValue === undefined ? key : { [key]: defaultValue }
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].get(keyObj, (items: any) => {
        const error = chrome.runtime.lastError
        if (error) return reject(error)
        resolve(items[key])
      })
    })
  },
  set: (key: string, value: any, storageArea: Storage) => {
    return new Promise((resolve, reject) => {
      chrome.storage[storageArea].set({ [key]: value }, () => {
        const error = chrome.runtime.lastError
        error ? reject(error) : resolve()
      })
    })
  },
}

/**
 * Basic hook for storage
 * @param {string} key
 * @param {*} initialValue
 * @param {'local'|'sync'} storageArea
 * @returns {[*, function(*= any): void, boolean, string]}
 */

export default function useChromeStorage(key, initialValue, storageArea) {
  const [INITIAL_VALUE] = useState(() => {
    return typeof initialValue === 'function' ? initialValue() : initialValue
  })
  const [STORAGE_AREA] = useState(storageArea)
  const [state, setState] = useState(INITIAL_VALUE)
  const [isPersistent, setIsPersistent] = useState(true)
  const [error, setError] = useState('')

  useEffect(() => {
    storage
      .get(key, INITIAL_VALUE, STORAGE_AREA)
      .then((res) => {
        setState(res)
        setIsPersistent(true)
        setError('')
      })
      .catch((error) => {
        setIsPersistent(false)
        setError(error)
      })
  }, [key, INITIAL_VALUE, STORAGE_AREA])

  const updateValue = useCallback(
    (newValue) => {
      const toStore = typeof newValue === 'function' ? newValue(state) : newValue
      storage
        .set(key, toStore, STORAGE_AREA)
        .then(() => {
          setIsPersistent(true)
          setError('')
        })
        .catch((error) => {
          // set newValue to local state because chrome.storage.onChanged won't be fired in error case
          setState(toStore)
          setIsPersistent(false)
          setError(error)
        })
    },
    [STORAGE_AREA, key, state]
  )

  useEffect(() => {
    const onChange = (changes, areaName) => {
      if (areaName === STORAGE_AREA && key in changes) {
        setState(changes[key].newValue)
        setIsPersistent(true)
        setError('')
      }
    }
    chrome.storage.onChanged.addListener(onChange)
    return () => {
      chrome.storage.onChanged.removeListener(onChange)
    }
  }, [key, STORAGE_AREA])

  return [state, updateValue, isPersistent, error]
}
