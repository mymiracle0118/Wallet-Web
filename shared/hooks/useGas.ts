import React, { useEffect, useState } from 'react'
import { useQuery } from '@tanstack/react-query'
import { ethers } from 'ethers'
import { getContractInterface } from '@portal/shared/utils/getContractInterface'
import { NetworkFactory } from '@portal/shared/factory/network.factory'
import { Asset, GasOption, NetworkToken } from '@portal/shared/utils/types'

const getERC20GasLimit = async (asset: Asset, provider: any): Promise<number> => {
  if (asset.type === 'erc721') {
    return 100000
  }
  const contract = new ethers.Contract(asset.contractAddress!, getContractInterface(asset.type), provider)
  const deploymentData = contract.interface.encodeDeploy([])
  const estimatedGasLimit = await provider.estimateGas({ data: deploymentData })
  return Number(ethers.utils.formatUnits(estimatedGasLimit, 'wei'))
}

export const useGas = (asset: NetworkToken, option?: GasOption) => {
  const [gasLimit, setGasLimit] = useState(21000)
  const [gasOption, setGasOption] = useState(option || GasOption.Aggressive)
  const getNetwork = NetworkFactory.selectByNetworkId(asset.networkName)
  const provider = getNetwork.provider
  const { data: feeData } = useQuery(
    ['gas-fee-data', asset],
    () => {
      return getNetwork.getFeeData()
    },
    {
      staleTime: 15000,
      refetchInterval: 15000,
    }
  )

  const { data: gasLimitData } = useQuery(
    ['gas-limit-data', asset, provider],
    () => getERC20GasLimit(asset, provider),
    {
      enabled: asset.tokenContractAddress !== undefined,
      staleTime: 60 * 60000,
    }
  )

  useEffect(() => {
    if (gasLimitData) {
      setGasLimit(gasLimitData)
    }
  }, [gasLimitData])

  const maxFeePerGas = React.useMemo(() => {
    if (!feeData || feeData.gasPrice === null) {
      return ethers.utils.parseUnits('0', 'gwei')
    }

    let baseGas = parseFloat(ethers.utils.formatUnits(feeData.gasPrice || 0, 'gwei'))
    const baseGasPlusMultiplier = (multiplier: number) => {
      return baseGas * multiplier
    }

    switch (gasOption) {
      case GasOption.Low:
        baseGas = baseGasPlusMultiplier(1.0)
        break
      case GasOption.Market:
        baseGas = baseGasPlusMultiplier(1.5)
        break
      case GasOption.Aggressive:
        baseGas = baseGasPlusMultiplier(2.0)
        break
    }

    return ethers.utils.parseUnits(baseGas.toFixed(8), 'gwei')
  }, [gasOption, feeData])

  const maxPriorityFeePerGas = React.useMemo(() => {
    return getNetwork.getMaxPriorityFeePerGas({ maxFeePerGas, gasOption })
  }, [gasOption, maxFeePerGas])

  const marketGas = React.useMemo(async () => {
    return await getNetwork.marketGas(maxFeePerGas)
  }, [gasOption, maxFeePerGas])

  const estimatedTransactionCost = React.useMemo(async () => {
    return maxFeePerGas && (await getNetwork.getEstimatedTransactionCost(maxFeePerGas, maxPriorityFeePerGas, gasLimit))
  }, [gasLimit, gasOption, maxFeePerGas])

  const estimatedTransactionTime = React.useMemo(() => {
    switch (gasOption) {
      case GasOption.Low:
        return '30 seconds'
      case GasOption.Market:
        return '30 seconds'
      case GasOption.Aggressive:
        return '15 seconds'
    }
  }, [gasOption])

  return {
    gasLimit,
    gasOption,
    setGasOption,
    marketGas,
    maxFeePerGas,
    maxPriorityFeePerGas,
    estimatedTransactionCost,
    estimatedTransactionTime,
  }
}
