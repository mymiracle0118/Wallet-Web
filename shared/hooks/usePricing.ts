import { useQuery } from '@tanstack/react-query'
import { utils } from 'ethers'
import React from 'react'

import { getSimplePrice } from '@portal/shared/services/coingecko'
import { NetworkToken } from '../utils/types'
import { useStore } from './useStore'

export const usePricing = (asset?: NetworkToken) => {
  const { currentTokenArrayWithBalance } = useStore()
  const { data } = useQuery(
    ['asset-pricing'],
    () =>
      getSimplePrice(
        [...new Set(Object.values(currentTokenArrayWithBalance).map((a) => a.coingeckoTokenId))].join(',')
      ),
    {
      refetchInterval: 40000,
      staleTime: 30000,
    }
  )

  const getMarketPrice = React.useCallback(
    (assetId: string) => {
      if (data && data[assetId]) {
        const total = Math.round((data[assetId].usd + Number.EPSILON) * 100) / 100
        return utils.commify(total)
      }

      return '0'
    },
    [data]
  )

  const getMarket24HourChange = React.useCallback(
    (assetId: string) => {
      if (data && data[assetId]) {
        return data[assetId].usd_24h_change
      }

      return 0
    },
    [data]
  )

  const getAssetValue = React.useCallback(
    (assetId: string, units: number) => {
      if (data && data[assetId]) {
        const value = Math.round((data[assetId].usd * units + Number.EPSILON) * 100) / 100
        return Math.max(value, 0)
      }

      return 0
    },
    [data]
  )

  const getAssetValueFormatted = React.useCallback(
    (assetId: string, units: number) => {
      return `$${getAssetValue(assetId, units)}`
    },
    [data]
  )
  const getAssetGasValueFormatted = React.useCallback(
    (assetId: string, units: number) => {
      let gasTotal = '0'
      if (data && data[assetId] && asset) {
        gasTotal = (data[assetId].usd * units).toFixed(parseInt(<string>asset?.decimal || '8'))
      }
      return `$${gasTotal}`
    },
    [data]
  )

  return { getMarketPrice, getMarket24HourChange, getAssetValue, getAssetValueFormatted, getAssetGasValueFormatted }
}
