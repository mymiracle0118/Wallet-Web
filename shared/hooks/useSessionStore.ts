import { encryptionDataSalt } from '@portal/portal-extension/src/app/services/config'
import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import { decryptData, encryptData } from '../services/EncryptionService'
import { chromeSessionStorage } from '../utils/chromeSyncStorage'

export type SessionStoreState = {
  clearStore: () => Promise<void>
  encryptedPassword: string | null
  getPassword: () => string | null
  setPassword: (password: string) => void
  isValidPassword: (password: string) => boolean
}

const initialState = {
  encryptedPassword: null,
}

export const useSessionStore = create<SessionStoreState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,
        clearStore: async () => {
          set(initialState)
        },
        getPassword: () => {
          const { encryptedPassword } = get()
          if (encryptedPassword) return decryptData(encryptedPassword as string, encryptionDataSalt)
          return null
        },
        isValidPassword: (password: string): boolean => {
          const exitingPassword = get().getPassword()
          return exitingPassword === password
        },
        setPassword: (password: string) => {
          const encryptedPassword: string = encryptData(password, encryptionDataSalt)
          set({ encryptedPassword })
        },
      }),
      {
        name: 'wallet-session-storage',
        getStorage: () => chromeSessionStorage,
      }
    ),
    { name: 'Star Key - Accounts' }
  )
)
