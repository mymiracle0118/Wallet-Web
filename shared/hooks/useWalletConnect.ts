import create from 'zustand'
import { devtools, persist } from 'zustand/middleware'
import type { IClientMeta, IWalletConnectSession } from '@walletconnect/legacy-types'
import WalletConnect from '@walletconnect/legacy-client'
import { ethers } from 'ethers'

import { chromeSyncStorage } from '@portal/shared/utils/chromeSyncStorage'
import { getNetworkProviderByChainId } from '@portal/shared/utils/getNetworkProvider'
import { useSettings } from './useSettings'

export type SessionRequest = {
  id: number
  method: string
  params: Array<{ peerId: string; peerMeta: IClientMeta; chainId: number }>
}

export type CallRequest = {
  id: number
  method: string
  params: Array<unknown>
}

export type WalletConnectState = {
  connector?: WalletConnect
  sessionRequest?: SessionRequest
  callRequest?: CallRequest
  session?: IWalletConnectSession
  selectedChainId: number
  createSession: (uri?: string) => void
  approveSession: (accounts: Array<string>) => void
  rejectSession: () => void
  approveCallRequest: (walletSigner: ethers.Wallet) => Promise<void>
  rejectCallRequest: () => void
  setSelectedChainId: (chainId: number) => void
  disconnect: () => void
  clear: () => void
}

const initialState = {
  selectedChainId: 1,
  connector: undefined,
  sessionRequest: undefined,
  callRequest: undefined,
  session: undefined,
}

const clientMeta = {
  description: 'Shuttle Wallet Developer App',
  url: 'https://walletconnect.org',
  icons: ['https://walletconnect.org/walletconnect-logo.png'],
  name: 'Shuttle Wallet',
}

export const useWalletConnect = create<WalletConnectState>()(
  devtools(
    persist(
      (set, get) => ({
        ...initialState,
        createSession: (uri?: string) => {
          let connector: WalletConnect

          const { session, selectedChainId } = get()
          if (session && session.connected) {
            connector = new WalletConnect({ session })
            if (selectedChainId !== 1) {
              connector.updateSession({ accounts: session.accounts, chainId: selectedChainId })
            }
          } else {
            connector = new WalletConnect({ uri, clientMeta })
          }

          set({ connector, session: connector.session })

          connector.on('session_request', (error, payload) => {
            set({ sessionRequest: payload })
          })

          connector.on('call_request', (error, payload) => {
            set({ callRequest: payload })
          })

          connector.on('disconnect', (error, payload) => {
            set(initialState)
          })
        },

        approveSession: (accounts: Array<string>) => {
          const { sessionRequest, connector, selectedChainId } = get()
          if (!sessionRequest) {
            throw new Error('No session request')
          }

          if (!connector) {
            throw new Error('No connector')
          }

          connector.approveSession({ accounts, chainId: selectedChainId })
          set({ sessionRequest: undefined, session: connector.session })
        },

        rejectSession: () => {
          const { connector, sessionRequest } = get()
          if (!sessionRequest) {
            throw new Error('No session request')
          }

          if (!connector) {
            throw new Error('No connector')
          }

          connector.rejectSession()
          set(initialState)
        },

        approveCallRequest: async (walletSigner: ethers.Wallet) => {
          try {
            const { connector, callRequest } = get()
            if (!callRequest) {
              throw new Error('No call request')
            }

            if (!connector) {
              throw new Error('No connector')
            }

            const verifyAddressMatch = (address: string) => {
              if (address.toLowerCase() !== walletSigner.address.toLowerCase()) {
                throw new Error('Address mismatch')
              }
            }

            let transaction: ethers.providers.TransactionRequest
            let dataToSign: string
            let result: string

            switch (callRequest.method) {
              case 'personal_sign': {
                dataToSign = callRequest.params[0] as string
                verifyAddressMatch(callRequest.params[1] as string)

                result = await walletSigner.signMessage(
                  ethers.utils.isHexString(dataToSign) ? ethers.utils.arrayify(dataToSign) : dataToSign
                )
                break
              }

              case 'eth_sign': {
                dataToSign = callRequest.params[1] as string
                verifyAddressMatch(callRequest.params[0] as string)

                result = await walletSigner.signMessage(
                  ethers.utils.isHexString(dataToSign) ? ethers.utils.arrayify(dataToSign) : dataToSign
                )
                break
              }

              case 'eth_signTransaction': {
                transaction = callRequest.params[0] as ethers.providers.TransactionRequest
                verifyAddressMatch(transaction.from as string)

                if ('gas' in transaction) {
                  // @ts-ignore
                  transaction.gasLimit = transaction.gas
                  // @ts-ignore
                  delete transaction.gas
                }

                result = await walletSigner.signTransaction(transaction)
                break
              }

              case 'eth_sendTransaction': {
                transaction = callRequest.params[0] as ethers.providers.TransactionRequest
                verifyAddressMatch(transaction.from as string)

                if ('gas' in transaction) {
                  // @ts-ignore
                  transaction.gasLimit = transaction.gas
                  // @ts-ignore
                  delete transaction.gas
                }

                result = (await walletSigner.sendTransaction(transaction)).hash
                break
              }

              case 'wallet_switchEthereumChain': {
                const { chainId } = callRequest.params[0] as { chainId: number }
                result = await walletSigner.signMessage(chainId.toString())
                break
              }

              default:
                throw new Error('Method not supported')
            }

            connector.approveRequest({ id: callRequest.id, result })
          } catch (error) {
            throw new Error(error.message)
          } finally {
            set({ callRequest: undefined })
          }
        },

        rejectCallRequest: () => {
          const { connector, callRequest } = get()
          if (!callRequest) {
            throw new Error('No  call request')
          }

          if (!connector) {
            throw new Error('No connector')
          }

          connector.rejectRequest({ id: callRequest.id, error: { message: 'User rejected request' } })

          set({ callRequest: undefined })
        },

        setSelectedChainId: (chainId: number) => {
          const { connector, session } = get()
          if (connector && session && session.connected) {
            connector.updateSession({ accounts: session.accounts, chainId: chainId })
          }

          set({ selectedChainId: chainId })
        },

        disconnect: async () => {
          const { connector } = get()
          if (!connector) {
            throw new Error('Connector is not initialized')
          }

          try {
            await connector.killSession()
          } catch (error) {
            console.error(error)
          } finally {
            set(initialState)

            const walletConnect = localStorage.getItem('walletconnect')
            if (walletConnect) {
              localStorage.removeItem('walletconnect')
            }
          }
        },

        clear: () => {
          set(initialState)
        },
      }),
      {
        name: 'wallet-connect-storage',
        getStorage: () => chromeSyncStorage,
      }
    ),
    { name: 'Shuttle - Wallet' }
  )
)

export const getWalletSigner = async (chainId: number, address: string, password: string) => {
  const account = Object.values(useSettings.getState().accounts)
    .map((a) => ({
      ...a,
      address: a.address.toLowerCase(),
    }))
    .find((a) => a.address === address.toLowerCase())

  if (!account) {
    throw new Error('Account not found')
  }

  try {
    const provider = getNetworkProviderByChainId(chainId)
    const wallet = await ethers.Wallet.fromEncryptedJson(account.encryptedWallet, password)
    return wallet.connect(provider)
  } catch (error) {
    throw new Error('Wrong password')
  }
}
