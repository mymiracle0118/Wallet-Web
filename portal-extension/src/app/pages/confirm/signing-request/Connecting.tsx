/* eslint-disable @typescript-eslint/no-unsafe-argument */

import React, { useEffect, useState } from 'react'
import SinglePageTitleLayout from 'layouts/single-page-layout/SinglePageLayout'
import connecting from 'assets/gif/connecting.gif'
import { CustomTypography, Icon } from 'components'
import ETH from 'assets/coins/ETH.svg'

import PasswordPromptModal from 'app/pages/wallet/PasswordPromptModal'
import { useWalletConnect, getWalletSigner } from '@portal/shared/hooks/useWalletConnect'
import { go, goBack } from 'lib/woozie'

const ConnectingToSignature = () => {
  const { session, callRequest, approveCallRequest } = useWalletConnect()

  const [openPasswordModal, setOpenPasswordModal] = useState<boolean>(false)

  useEffect(() => {
    if (!callRequest) {
      go(-1)
    } else {
      setOpenPasswordModal(true)
    }
  }, [callRequest])

  const handlePromptPassword = async (password: string) => {
    try {
      setOpenPasswordModal(false)
      const walletSigner = await getWalletSigner(session?.chainId as number, session?.accounts[0] as string, password)
      await approveCallRequest(walletSigner)
      go(-2)
    } catch (error) {
      setOpenPasswordModal(true)
    }
  }

  const handleCancelPasswordPrompt = () => {
    setOpenPasswordModal(false)
    goBack()
  }

  return (
    <SinglePageTitleLayout showWalletConnectMenu disableBack>
      <div className="flex items-center justify-center gap-4 mt-48">
        <Icon icon={<ETH />} size="large" />
        <img src={connecting} alt="my-gif" className="w-32" />
        <img className="h-9 w-9 rounded-full" alt="token-thumbnail" src={session?.peerMeta?.icons[0]} />
      </div>
      <CustomTypography className="text-xl text-center mt-8 !text-secondary font-bold">Connecting...</CustomTypography>
      <PasswordPromptModal
        modalState={openPasswordModal}
        closePromptModal={handleCancelPasswordPrompt}
        onPromptPassword={handlePromptPassword}
      />
    </SinglePageTitleLayout>
  )
}

export default ConnectingToSignature
