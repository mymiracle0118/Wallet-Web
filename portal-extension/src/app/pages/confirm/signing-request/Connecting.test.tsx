import React from 'react'
import { render } from '@testing-library/react'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { I18nextProvider } from 'react-i18next'
import i18n from '@portal/shared/i18n/config'

import * as useWalletConnectHooks from '@portal/shared/hooks/useWalletConnect'
import Connecting from './Connecting'

const mockWalletConnectHook = {
  session: {
    peerMeta: {
      name: 'WalletConnect',
      icons: ['https://walletconnect.org/walletconnect-logo.png'],
      url: 'https://walletconnect.org',
    },
  },
}

const queryClient = new QueryClient()
const wrapper = (
  <QueryClientProvider client={queryClient}>
    <I18nextProvider i18n={i18n}>
      <Connecting />
    </I18nextProvider>
  </QueryClientProvider>
)

describe('<Connecting />', () => {
  beforeEach(() => {
    jest.spyOn(useWalletConnectHooks, 'useWalletConnect').mockImplementation(() => mockWalletConnectHook)
  })

  test('renders correctly', () => {
    const { asFragment } = render(wrapper)
    expect(asFragment()).toMatchSnapshot()
  })
})
