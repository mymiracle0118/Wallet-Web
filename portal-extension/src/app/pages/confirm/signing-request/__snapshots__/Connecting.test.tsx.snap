// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Connecting /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dark:bg-surface-dark bg-custom-white flex flex-col h-[600px] dark:text-custom-white text-custom-black"
  >
    <div
      class="flex flex-row relative justify-between items-center h-16"
    >
      <button
        class="profile-btn flex items-center mr-1 text-inherit font-bold hover:bg-custom-grey10 dark:hover:bg-custom-white10 p-3 rounded-md"
        type="button"
      >
        @
        <chevron-right.svg
          classname="stroke-custom-black dark:stroke-custom-white"
        />
      </button>
    </div>
    <div
      class="w-full h-full overflow-auto"
    >
      <div
        class="flex items-center justify-center gap-4 mt-48"
      >
        <svg
          class="inline-block select-none shrink-0 fill-custom-black dark:fill-custom-white text-4xl"
          style="height: 1em; width: 1em;"
        >
          <eth.svg />
        </svg>
        <img
          alt="my-gif"
          class="w-32"
          src="connecting.gif"
        />
        <img
          alt="token-thumbnail"
          class="h-9 w-9 rounded-full"
          src="https://walletconnect.org/walletconnect-logo.png"
        />
      </div>
      <div
        class="text-xl text-center mt-8 !text-secondary font-bold text-custom-black dark:text-custom-white"
      >
        Connecting...
      </div>
      <div
        style="position: absolute; width: 1px; height: 1px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
      />
    </div>
  </div>
</DocumentFragment>
`;
